Adds ROCKET (transform only) to sktime.

angus924 opened the pull request:
As requested, this adds [ROCKET](https://arxiv.org/abs/1910.13051) to sktime.

**Notes:**

* This is the ROCKET transform only; the transformed features should be used with a classifier (we use sklearn/RidgeClassifierCV in our experiments for 'smaller' datasets).  I can add a separate sktime classifier class to link the transform and the sklearn classifier.
* This implementation can handle univaraite and multivariate time series, but not (yet) variable-length time series directly.
* As ROCKET relies on Numba, and Numba only understands NumPy (not pandas), this code uses an unavoidable 'hack', in that the input DataFrame is (and must be) converted into a NumPy array internally.  (Almost certainly, this conversion will break for variable-length time series.)
* The implementation is preliminary, in particular in relation to multivariate time series, which we are still investigating.
* This should produce equivalent results (including timings) to our current published implementation (it is fundamentally very similar), but I have not yet tested this extensively.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+import numpy as np
+import pandas as pd
+
+from .base import BaseTransformer
```
```suggestion
from sktime.base import BaseTransformer
```

angus924 replied:
The codebase seems to use mostly absolute imports.  However, the contributing and extension guidelines both stipulate relative imports.

In any case, should this be: `from sktime.transformers.base import BaseTransformer`?

mloning replied:
Yes, thanks, sorry, need to update extension guidelines! We now prefer absolute imports ...

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+        return self
+
+    def fit_transform(self, X, y = None):
+        return self.fit(X).transform(X)
```
We can remove the `fit_transform` method, that comes automatically with the [`BaseTransformer`](https://github.com/scikit-learn/scikit-learn/blob/1495f6924/sklearn/base.py#L526).

angus924 replied:
Do you mean removing `fit_transform` altogether from the base class (`BaseTransformer`)?

`fit` (and, by extension `fit_transform`) are irrelevant for ROCKET.

mloning replied:
Yes! Just to clarify in case it's not clear: the TransformerMixin form sklearn (and by extension our BaseTransformer) has a `fit_transform` method already which calls `.fit(X).transform(X)` which is inherited by your transformer, so no need to add (or overwrite) it here. Even if fit doesn't do anything, we should keep it here to have a consistent interface.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+        return pd.DataFrame(_apply_kernels(self._to_numpy(X), self.kernels))
+
+    @staticmethod
+    def _to_numpy(X, a = None, b = None):
```
We do have a tabularisation function, which essentially does the same, how does your method represent multivariate time series data? I'd prefer if we move this function to the [`sktime.utils.data_container`](https://github.com/alan-turing-institute/sktime/blob/master/sktime/utils/data_container.py) file as we may want to add this functionality to the data container eventually.

angus924 replied:
-oIt looks like `tabularize` 'flattens' the input into 2d.  This `_to_numpy` function produces an ndarray with three dimensions: (*number_of_examples*, *number_of_channels*, *length_of_time_series*).  (The `a` and `b` parameters are in order to be able to do the conversion in a rolling fashion in order to, in turn, avoid duplicating a big dataset in memory.)

This is basically the input format for our original implementation as well (extended to multivariate).  The problem, of course, comes with variable-length time series, for which I haven't really settled on a solution.  My favoured one is basically to mimic the backend of a scipy sparse matrix, store two sets of values: all the data in a single 1d numpy array, plus a set of lengths / cutpoints.

mloning replied:
Alright, let's keep it, I'd still prefer the `_to_numpy` method to go in the utils file. We've been reconsidering our data representation lately, let me know if you're interested, by the way are you in our slack?

angus924 replied:
No worries, I'll move the function.

Definitely interested, if only to learn more (I don't pretend to know much about anything useful).  Not in the slack.

(I've been meaning to try [awkward-array](https://github.com/scikit-hep/awkward-array), as it looks like it might solve some of the issues with handling this kind of data...)

mloning replied:
Alright, sent an invitation email for our slack!

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+    a1 = 0 # for weights
+    a2 = 0 # for channel_indices
+
+    for i in range(num_kernels):
```
Have you had a look at parallelisation with joblib? We're not sure if introducing numba as an additional dependency is a good idea at the moment.

angus924 replied:
Obviously, in terms of the project, this decision is up to you.  However, if this implementation of ROCKET doesn't use numba (or, maybe, something else roughly equivalent), it really can't be used for benchmarking.

Unless you have a fairly fancy CPU, or access to a CPU cluster, parallelisation is likely only going to be a secondary benefit of using numba.  The main benefit of using numba is in compiling the functions that are doing all the work.  (In our experiements, we basically didn't use parallelisation: each experiment / dataset / run was performed on a single CPU core / thread.)

Given the nature of the computations required for ROCKET, compilation is really fundamental to making a python implementation work.  (In this implementation here, numba performs compilation is invisibly at import, and provides a massive, massive speedup over uncompiled python / numpy syntax.)

More generally, although my experience with numba is relatively limited, I strongly recommend it.

mloning replied:
Ok, thanks, yes I've used numba before, as I've said, not really sure at this point, for the distance metrics we chose to write them in Cython and then parallelise with joblib.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+    """
+
+    def __init__(self,  input_length, num_kernels, num_channels = 1):
+        self.input_length = input_length
```
Why do we have to specify the input length at construction? Can we not infer this from the data during fitting?

angus924 replied:
Input length is basically just used to work out an appropriate maximum kernel dilation.  Maybe `reference_length` would be more appropriate, but would require more explanation.

It could be inferred from the input, and for time series all of the same length, in probably most cases, this would be fine.

It becomes tricker for variable-length time series.  Here, there isn't necessarily one input length or reference length which makes sense.  You could set this as the length of the longest time series, or the mean or median length, etc.  (This can be a real issue for some datasets where the length of the longest time series is orders of magnitude longer than the mean.)

I guess the training data could be an optional input to the constructor or `fit`, just to infer the dimensions.

mloning replied:
I like your last suggestion, I think that makes most sense: use `fit` to infer heuristics or variables that depend on the data which are then used in transform. Adding a `check_is_fitted` would be great too.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+    ----------
+    input_length : int, length of input time series
+    num_kernels  : int, number of random convolutional kernels
+    num_channels : int, number of channels for multivariate time series (default 1)
```
What do you mean by channels? Variables in the multivariate setting?

angus924 replied:
I think `channel == dimension`, so each column in a sktime pandas DataFrame would be a channel in this terminology (coming from cnns).  Happy to change.

mloning replied:
Alright, we've started writing a little [glossary](https://github.com/alan-turing-institute/sktime/wiki/Glossary) to keep terminology consistent: so variables/channels/dimensions are referred to as columns.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+        return np.stack(X.iloc[a:b].applymap(lambda cell : cell.to_numpy()).apply(lambda row : np.stack(row), axis = 1).to_numpy())
+
+@njit("Tuple((float32[:],int32[:],float32[:],int32[:],int32[:],int32[:],int32[:]))(int64,int64,int64)")
+def _generate_kernels(input_length, num_kernels, num_channels):
```
Could we perhaps add this as a method to the transformer, so that we can use a random state similar to sklearn for reproducibility?

angus924 replied:
`_generate_kernels` can be moved into the class easiliy (the other two functions are trickier for obscure reasons).  Either way, it should be possible to use a seed / random state.

mloning replied:
Ok, as long as we can have a random state in the constructor which is propagated to the functions, please use sklearn's `check_random_state` function :-)

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,178 @@
+        -------
+        pandas DataFrame, transformed features
+        """
+        return pd.DataFrame(_apply_kernels(self._to_numpy(X), self.kernels))
```
We should add some appropriate input checks from [here](https://github.com/alan-turing-institute/sktime/blob/master/sktime/utils/validation/supervised.py), i.e. `validate_X` and possibly also `check_X_is_univariate`?

angus924 replied:
I'll do that.

`check_X_is_univariate` shouldn't matter, as it should work just as well whether the input DataFrame has 1, 5, or 500 columns / dimensions.

angus924 commented:
Thanks very much for taking the time to look over the code.

I'll have a go at adding the unit test you suggest.

In relation to numba, just further to my response above, I'd say that compilation via numba (or something roughly equivalent) is essential to making a python implementation of ROCKET work, quite apart from parallelism.  In addition, I would strongly recommend considering numba more generally, as it is pretty impressive what it can do in terms of compilation / speed and parallelism writing pure python / numpy syntax (although you are a bit restricted in terms of what syntax it understands).  It could be useful in this project more broadly (but this is pure speculation on my part).

mloning commented:
Regarding numba discussion, see #84

mloning commented on sktime/transformers/tests/test_Rocket.py:
```diff
@@ -0,0 +1,30 @@
+from sktime.transformers.rocket import Rocket
+from sklearn.linear_model import RidgeClassifierCV
+
+def test_rocket_on_gunpoint():
```
Thanks for adding the test! 
 * It'd be great to have some `assert` statements in the test (e.g. using the `np.testing` modules) checking dimension/shapes of output and so on, 
* I wouldn't do anything too computationally expensive in unit tests, so no need to run grid-search cv or many kernels.

angus924 replied:
I'll add some checks.

I think the `test_rocket_on_gunpoint(...)` function should be ok in terms of runtime: it runs in about 3 seconds on my 5-year-old laptop (if this is too much, I can reduce the number of kernels).

`RidgeClassifierCV` by default uses generalized cross-validation (a kind of estimated loocv), so the time cost of performing the cross-validation should be basically imperceptible, at least on small datasets.  Here the classifier runtime (training + test) is about 70 milliseconds including cross-validation.

angus924 commented:
Is there anything else needed for this pull request at this stage?

I can see that the builds are failing, I assume because numba isn't in the `dev` branch dependencies.  Would you like me to fix / change that, or just leave it for now?

Thanks very much.

mloning commented:
Yes, it's failing because of the missing numba dependency, happy if you add this to the PR.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,239 @@
+
+    weights, lengths, biases, dilations, paddings, num_channel_indices, channel_indices = kernels
+
+    num_examples, num_channels, _ = X.shape
```
```suggestion
    n_instances, n_columns, n_timepoints = X.shape
```

Would be great to use our agreed terminology here (if my interpretation of your code is correct)

mloning replied:
Tests are passing now, so that should be the last bit before we can merge it! We're currently writing down a list of methods and their contributors, so I may ask you to add yourself with your method to the table.

mloning commented:
I've thought about the additional numba dependency, I believe the main problem will be if someone uses the ROCKET transform in a pipeline together with GridSearchCV to tune one of ROCKET's hyper-parameters, from my experience joblib and numba don't work well together. 

That would be something to try out, if it breaks, I'd suggest to put ROCKET into a [companion repo for sktime](https://github.com/sktime) for now until we figure out how to make them work together and to reduce external dependencies for the core package (numba).

angus924 commented:
Re GridSearchCV:

* For ROCKET, the only thing to possibly 'tune' is `num_kernels`.  Broadly speaking, you shouldn't normally have to worry about this.  The default value, 10,000, shouldn't need to change (bigger is usually better, but there's not much point making it bigger than 10,000).
* Unrelated to numba / joblib, I couldn't get GridSearchCV working at all without some changes:
  * adding a 'dummy' Y parameter to the ROCKET `fit` and `transform` methods (sklearn/GridSearchCV insists on passing Y values through the estimator); and
  * deleting `fit_params` from the call to init in sktime/GridSearchCV (I think this parameter is gone from more recent versions of sklearn).
* However, once it was working (with the above changes), there didn't seem to be any issues between numba + joblib (everything was working as expected for n_jobs = -1, n_jobs > 1, etc., on my machine).

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,239 @@
+import numpy as np
+import pandas as pd
+
+from sklearn.utils.validation import check_is_fitted, check_random_state
```
```suggestion
from sklearn.utils.validation import check_random_state
from sktime.utils.validation import check_is_fitted
```

mloning replied:
One more change as we just changed this on `dev`, should be able to simply accept the suggestions.

mloning commented:
Okay, thanks for checking, in that case I'm comfortable to merge the PR. I've just made one last suggestion to replace the import above.

mloning commented:
Please also go through this: 

# PR checklist
- [x] I have added unit tests and made sure they pass locally. 
- [ ] I have updated the existing example notebooks or provided a new one to showcase how my algorithm works.
- [ ] I have updated sktime's [estimator overview](https://github.com/alan-turing-institute/sktime/blob/master/ESTIMATOR_OVERVIEW.md) and I'm committed to maintain my contribution and provide bug fixes if necessary.

Sorry for all the delay on the PR, you've opened the PR in a time when we went through some discussions on how to best integrate new contributions, etc.

angus924 commented:
Apologies for the long delay; it's been a very busy time.

I've fixed the import, and added an example notebook, and upated the estimator overview.

Thanks for your patience.

# PR checklist
* [x]  I have added unit tests and made sure they pass locally.
* [x]  I have updated the existing example notebooks or provided a new one to showcase how my algorithm works.
* [x]  I have updated sktime's [estimator overview](https://github.com/alan-turing-institute/sktime/blob/master/ESTIMATOR_OVERVIEW.md) and I'm committed to maintain my contribution and provide bug fixes if necessary.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,240 @@
+        self._is_fitted = True
+        return self
+
+    def transform(self, X, normalise = True):
```
```suggestion
    def transform(self, X, y=None, normalise=True):
```
Also could `normalise` be set in the constructor?

mloning replied:
Please make sure `y=None` is in the transform and fit signature!

angus924 replied:
Done: `normalise` moved to the constructor, and `y` added to `fit` and `transform`.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,240 @@
+    seed        : int, random seed (optional, default None)
+    """
+
+    def __init__(self, num_kernels = 10_000, seed = None):
```
Please rename `seed` to `random_state` to have common names!

angus924 replied:
Done, but see comment below.

mloning commented on sktime/transformers/rocket.py:
```diff
@@ -0,0 +1,240 @@
+def _generate_kernels(n_timepoints, num_kernels, n_columns, seed):
+
+    if seed is not None:
+        np.random.seed(seed)
```
Could we not use `check_random_state(random_state)` which returns a random state instance which you can then use to call the random functions, e.g.
```python
rng = check_random_state(random_state)
rng.choice(...)
```

angus924 replied:
The NumPy `RandomState` (produced by `check_random_state`) can't be used inside Numba-compiled code.  Numba uses its own random state.   I've changed the code to accept an integer argument (seed), and ignore everything else.  This way it's compatible with the `random_state` parameter in other sktime code (i.e., won't throw an exception, etc.), but you are limited to using an integer seed.  I think that's the best I can do at this point.

mloning replied:
Okay, let's leave it as it is, from what I've read they're trying to make numpy and numba more consistent but still working on it.

mloning commented on sktime/transformers/tests/test_Rocket.py:
```diff
@@ -0,0 +1,40 @@
+
+    # predict (alternatively: 'classifier.score(X_test_transform, Y_test)')
+    predictions = classifier.predict(X_test_transform)
+    accuracy = (predictions == Y_test).mean()
```
Please use the sklearn function for this: `accuracy_score`

angus924 replied:
Done.

mloning commented on sktime/transformers/tests/test_Rocket.py:
```diff
@@ -0,0 +1,40 @@
+    accuracy = (predictions == Y_test).mean()
+
+    # test predictions (on Gunpoint, should be 100% accurate)
+    np.testing.assert_array_equal(predictions, Y_test)
```
Isn't that the same as
```suggestion
    assert accuracy == 1
```
?

angus924 replied:
Yes; changed per your suggestion.

mloning commented on sktime/utils/data_container.py:
```diff
@@ -1,230 +1,245 @@
+
+    return time_index
+
+def dataframe_to_numpy(X, a = None, b = None):
```
Thanks for the utility function, could we rename it to `nested_to_3d_numpy`? This is very useful, as it also provides an interface to the tslearn package 👍

angus924 replied:
Renamed to `nested_to_3d_numpy`.

angus924 commented:
Thanks for the very prompt review.

I've made the requested changes (with the caveat re `random_state`, above), and updated the notebook with a pipeline example.

Thanks very much.

mloning commented:
Good stuff! Good to merge from my side, anything else from your side?

angus924 commented:
Thanks again for taking the time to look over everything (and doing it so quickly).

Nothing from my side.  All good.

TonyBagnall commented:
great job Angus, sorry for the silence, very busy. I'll get someone over her to give it a road test and let you know the results.
